File: cses/07_mathematics/1081_common-divisors/solution.py
import sys


def main():
    # Đọc toàn bộ dữ liệu; M là giá trị lớn nhất xuất hiện trong mảng.
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    M = 0
    ints = []
    for i in range(1, n + 1):
        v = int(data[i])
        ints.append(v)
        if v > M:
            M = v

    # cnt[v] = số phần tử của mảng có giá trị đúng bằng v.
    cnt = [0] * (M + 1)
    for v in ints:
        cnt[v] += 1

    # Harmonic sieve: duyệt ước d từ lớn xuống nhỏ, số phần tử chia hết cho d là
    # sum(cnt[d::d]) (cộng cnt tại d, 2d, 3d, ...) tính ở tốc độ C.
    # d đầu tiên có >= 2 bội chính là gcd lớn nhất của một cặp -> đáp án.
    for d in range(M, 0, -1):
        if sum(cnt[d::d]) >= 2:
            sys.stdout.write(str(d))
            return

    # Vì n >= 2 nên luôn tìm thấy ở trên; nhánh này chỉ để phòng hờ.
    sys.stdout.write("1")


main()

File: cses/07_mathematics/1081_common-divisors/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // Đọc mảng và ghi nhận giá trị lớn nhất M.
    int n;
    cin >> n;
    vector<int> values(static_cast<size_t>(n));
    int maximum = 0;
    for (int& value : values) {
        cin >> value;
        maximum = max(maximum, value);
    }

    // frequency[v] = số phần tử của mảng có giá trị đúng bằng v.
    vector<int> frequency(static_cast<size_t>(maximum) + 1, 0);
    for (int value : values) {
        ++frequency[static_cast<size_t>(value)];
    }

    // Duyệt ước d từ lớn xuống nhỏ (harmonic sieve). Với mỗi d, đếm số phần tử
    // chia hết cho d bằng cách cộng frequency tại các bội d, 2d, 3d, ...
    // d đầu tiên có >= 2 bội chính là gcd lớn nhất của một cặp -> đáp án.
    for (int divisor = maximum; divisor >= 1; --divisor) {
        int divisible_count = 0;
        for (int multiple = divisor; multiple <= maximum; multiple += divisor) {
            divisible_count += frequency[static_cast<size_t>(multiple)];
            if (divisible_count >= 2) {
                cout << divisor << '\n';
                return 0;
            }
        }
    }
}
